app/src/main/java/com/babylon/wallet/android/domain/usecases/p2plink/EstablishP2PLinkConnectionUseCase.kt

```diff
@@ -0,0 +1,123 @@
+import kotlinx.serialization.json.Json
+import rdx.works.core.decodeHex
+import rdx.works.core.mapWhen
+import rdx.works.core.toHexString
+import rdx.works.peerdroid.data.PeerdroidLink
+import rdx.works.profile.data.model.apppreferences.P2PLink
+import rdx.works.profile.ret.crypto.PrivateKey
+import timber.log.Timber
+import javax.inject.Inject
+
+class EstablishP2PLinkConnectionUseCase @Inject constructor(
+    private val peerdroidLink: PeerdroidLink,
+    private val p2PLinksRepository: P2PLinksRepository
+) {
+
+    suspend fun update(payload: LinkConnectionPayload): Result<Unit> {
+        val existingLink = payload.existingP2PLink ?: return Result.failure(
+            IllegalStateException("The existing p2p link can't be null at this point")
+        )
+
+        return addConnection(
+            payload = payload,
+            saveP2PLink = { p2pLinks, privateKey ->
+                val updatedP2PLink = payload.toP2PLink(
+                    name = existingLink.displayName,
+                    privateKey = privateKey
+                )
+                p2PLinksRepository.save(
+                    p2pLinks.mapWhen(
+                        predicate = { it.isSame(updatedP2PLink) },
+                        mutation = { updatedP2PLink }
+                    )
+                )
+            }
+        )
+    }
+
+    suspend fun add(payload: LinkConnectionPayload, name: String): Result<Unit> {
+        return addConnection(
+            payload = payload,
+            saveP2PLink = { p2pLinks, privateKey ->
+                val newP2PLink = payload.toP2PLink(
+                    name = name,
+                    privateKey = privateKey
+                )
+                p2PLinksRepository.save(p2pLinks + newP2PLink)
+            }
+        )
+    }
+
+    private suspend fun addConnection(
+        payload: LinkConnectionPayload,
+        saveP2PLink: suspend (p2pLinks: List<P2PLink>, privateKey: String) -> Unit
+    ): Result<Unit> {
+        val encryptionKey = parseEncryptionKeyFromConnectionPassword(
+            connectionPassword = payload.password
+        ) ?: return Result.failure(IllegalArgumentException("Failed to parse encryption key from connection password"))
+
+        return peerdroidLink.addConnection(
+            encryptionKey = encryptionKey,
+            connectionListener = object : PeerdroidLink.ConnectionListener {
+
+                override suspend fun completeLinking(connectionId: String): Result<Unit> {
+                    val p2pLinks = p2PLinksRepository.getP2PLinks()
+
+                    val walletClientKeyBytes = p2pLinks.findBy(payload.publicKey)?.walletPrivateKey
+                        ?.decodeHex() ?: PrivateKey.EddsaEd25519.newRandom().toByteArray()
+                    val walletClientKey = PrivateKey.EddsaEd25519.newFromPrivateKeyBytes(walletClientKeyBytes)
+
+                    val linkClientInteractionResponse = getLinkClientInteractionResponse(
```

Comment:
minor: maybe `buildLinkClientInteractionResponse` is more suitable here? Since when I saw "get" I thought it was a response from linking